testdata/p4_16_samples/match-on-exprs2-bmv2.p4

```diff
@@ -0,0 +1,103 @@
+/*
+Copyright 2019 Cisco Systems, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include <core.p4>
+#include <v1model.p4>
+
+typedef bit<48>  EthernetAddress;
+
+header ethernet_t {
+    EthernetAddress dstAddr;
+    EthernetAddress srcAddr;
+    bit<16>         etherType;
+}
+
+struct headers_t {
+    ethernet_t    ethernet;
+}
+
+struct metadata_t {
+}
+
+parser parserImpl(packet_in packet,
+                  out headers_t hdr,
+                  inout metadata_t meta,
+                  inout standard_metadata_t stdmeta)
+{
+    state start {
+        packet.extract(hdr.ethernet);
+        transition accept;
+    }
+}
+
+control verifyChecksum(inout headers_t hdr, inout metadata_t meta) {
+    apply { }
+}
+
+control ingressImpl(inout headers_t hdr,
+                    inout metadata_t meta,
+                    inout standard_metadata_t stdmeta)
+{
+    action my_drop() {
+        mark_to_drop(stdmeta);
+    }
+    action foo(bit<9> out_port) {
+        hdr.ethernet.dstAddr[22:18] = hdr.ethernet.srcAddr[5:1];
+        stdmeta.egress_spec = out_port;
+    }
+    table t1 {
+        key = {
+            hdr.ethernet.srcAddr[22:18] : exact;
+            hdr.ethernet.dstAddr & 0x010101010101 : exact;
+            hdr.ethernet.etherType - 10 : exact @name("etherType_less_10");
```

Comment:
The P4Info message is exactly what I would expect. I guess ideally the `@name` annotation on `hdr.ethernet.etherType - 10` would not be required, and name `"hdr.ethernet.etherType - 10"` would be auto-generated by default.
Given that in P4Runtime, match field names are just strings which are mapped to ids (with ids being used in P4Runtime messages), the presence of unusual characters  in the names, such as whitespaces, should not be an issue for any implementation. This is unlike the bmv2 JSON and the tools which consume it, which require the match field to be a plain field reference. But the bmv2 CLI doesn't really use these names when receiving commands from the user, instead it treats the match key as a vector of fields (and so does the bmv2 Thrift API). It seems that on the other hand STF uses names to identify match fields.

Reply:
By the way @jafingerhut, I think once this is settle, this information would be a great addition to your https://github.com/p4lang/p4runtime/blob/master/docs/v1/guidance-for-generating-p4info.md document